crates/ruma-client-api/src/rendezvous/create_rendezvous_session.rs

```diff
@@ -0,0 +1,235 @@
+        const METADATA: Metadata = METADATA;
+
+        fn try_from_http_request<B, S>(
+            request: http::Request<B>,
+            _path_args: &[S],
+        ) -> Result<Self, ruma_common::api::error::FromHttpRequestError>
+        where
+            B: AsRef<[u8]>,
+            S: AsRef<str>,
+        {
+            let content_type = request
+                .headers()
+                .get(CONTENT_TYPE)
+                .ok_or(HeaderDeserializationError::MissingHeader(CONTENT_TYPE.to_string()))?;
+
+            if content_type.as_bytes() != b"text/plain" {
+                Err(serde_json::Error::custom(
+                    "Invalid content type. The content type must be `text/plain`.",
+                )
+                .into())
+            } else {
+                let body = request.into_body().as_ref().to_vec();
+                let content = String::from_utf8(body).map_err(|e| {
+                    serde_json::Error::custom(format!("Content isn't a valid UTF-8 string {e:?}"))
+                })?;
+
+                Ok(Self { content })
+            }
+        }
+    }
+
+    impl Request {
+        /// Creates a new `Request` with the given content.
+        pub fn new(content: String) -> Self {
+            Self { content }
+        }
+    }
+
+    impl Default for Request {
+        fn default() -> Self {
+            Self::new("".to_owned())
+        }
+    }
+
+    /// Response type for the `POST` `rendezvous` endpoint.
+    #[derive(Debug, Clone)]
+    #[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+    pub struct Response {
+        /// The absolute URL of the rendezvous session.
+        pub url: Url,
+
+        /// ETag for the current payload at the rendezvous session as
+        /// per [RFC7232](https://httpwg.org/specs/rfc7232.html#header.etag).
+        pub etag: String,
+
+        /// The expiry time of the rendezvous as per
+        /// [RFC7234](https://httpwg.org/specs/rfc7234.html#header.expires).
+        pub expires: SystemTime,
+
+        /// The last modified date of the payload as
+        /// per [RFC7232](https://httpwg.org/specs/rfc7232.html#header.last-modified)
+        pub last_modified: SystemTime,
+    }
+
+    #[derive(Serialize, Deserialize)]
+    struct ResponseBody {
+        url: Url,
+    }
+
+    #[cfg(feature = "client")]
+    impl ruma_common::api::IncomingResponse for Response {
+        type EndpointError = crate::Error;
+
+        fn try_from_http_response<T: AsRef<[u8]>>(
+            response: http::Response<T>,
+        ) -> Result<Self, FromHttpResponseError<Self::EndpointError>> {
+            if response.status().as_u16() >= 400 {
+                return Err(FromHttpResponseError::Server(
+                    Self::EndpointError::from_http_response(response),
+                ));
+            }
+
+            let get_date = |header: HeaderName| -> Result<SystemTime, FromHttpResponseError<Self::EndpointError>> {
```

Comment:
We could simplify your code by having common helpers to convert headers to/from HTTP dates, which would be great.

That would be a breaking change though, as it would mean making the error from `RetryAfter` conversion less specific.

Reply:
It did it in #1802

Reply:
Thanks, I rebased and adapted things in https://github.com/ruma/ruma/pull/1800/commits/9fe6820c0cf7a3bbe143af831e009eaecd6d45ad.